compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -64,6 +65,105 @@ static Value linearizeIndex(OpBuilder &builder, Value offset,
       .getResult();
 }
 
+/// Given a set of base transfer |indices|, |offsets| for the batch/outer
+/// dimensions, and distributed warp and thread indices, computes the indices
+/// of the distributed transfer operation based on the |vectorLayout|.
+static SmallVector<Value> getTransferIndicesFromNestedLayout(
+    OpBuilder &b, ValueRange indices, ArrayRef<int64_t> offsets,
+    NestedLayoutAttr vectorLayout, AffineMap permutationMap,
+    ArrayRef<Value> warpIndices, ArrayRef<Value> threadIndices) {
+  auto isBroadcast = [](AffineExpr expr) {
+    if (auto constExpr = dyn_cast<AffineConstantExpr>(expr))
+      return constExpr.getValue() == 0;
+    return false;
+  };
+  int64_t rank = vectorLayout.getBatchOrder().size();
+  // Permute the batch and outer vector offsets to match the order of
+  // the vector dimensions using the inverse of the batch/offset order.
+  SmallVector<int64_t> batchOffsets =
+      applyPermutation(ArrayRef<int64_t>(offsets.begin(), rank),
+                       invertPermutationVector(vectorLayout.getBatchOrder()));
+  SmallVector<int64_t> outerVectorOffsets =
+      applyPermutation(ArrayRef<int64_t>(offsets.begin() + rank, rank),
+                       invertPermutationVector(vectorLayout.getOuterOrder()));
+
+  SmallVector<Value> slicedIndices(indices.begin(), indices.end());
+  for (const auto &[i, dim] : llvm::enumerate(permutationMap.getResults())) {
+    // Broadcasted dimension offsets can be used as-is; the read index is
+    // invariant of the thread in such cases (and illegal for writes).
+    if (isBroadcast(dim)) {
+      continue;
+    }
+    unsigned pos = cast<AffineDimExpr>(dim).getPosition();
+    SmallVector<OpFoldResult> ids = {
+        warpIndices[i], b.getIndexAttr(batchOffsets[i]),
+        b.getIndexAttr(outerVectorOffsets[i]), threadIndices[i]};
+    // The order in which a vector dimension is "tiled" is
+    // subgroups -> batches -> outer vectors -> threads -> elements
+    SmallVector<int64_t> sizes = {vectorLayout.getSubgroupsPerWorkgroup()[i],
+                                  vectorLayout.getBatchesPerSubgroup()[i],
+                                  vectorLayout.getOutersPerBatch()[i],
+                                  vectorLayout.getThreadsPerOuter()[i]};
```

Comment:
We should probably add a helper to the layout for this (Not required as part of this PR, just thinking out loud.)